### ISSUE TITLE
Effects rendered are duplicated even coming from weapons

### ISSUE THREAD

**Issue by ensiform** [JACoders/OpenJK#908]
Reporting a bug? Please make sure you've given the following information - thanks!

**Operating system and version:**
All

**Is this for single player or multiplayer?**
Both (but it's JKA specific)

**Description of the bug (and if possible, steps to reproduce the bug):**
Many projectiles have effects duplicated up to even 7 times in some cases like the repeater primary fire attack.

**What did you expect to happen instead?**
Only the listed effect primitives should render as expected from the effect file.

**Comment by Grab11** [comment]
All weapons effects(including SP weapons like Noghri stick) minus Rocket Launcher are "spawned" 7 times instead of once. DEMP2 projectiles are spawned twice. This affects projectiles only - not explosions. This is not happening in JK2. 
Screenshots:
http://i.imgur.com/HfQA6fr.jpg
http://i.imgur.com/LJU7Buw.jpg
http://i.imgur.com/xULha0a.png

**Comment by ensiform** [comment]
@xycaleth any ideas? Its not specific to OpenJK. Just JKA in general.

**Comment by xycaleth** [comment]
No idea. Good idea with the wireframe to see what was happening though :) Does it still duplicate it that many times at lower frame rates? Try using com maxfps to cap fps.

**Comment by Grab11** [comment]
@xycaleth Nice shot
125 FPS: http://i.imgur.com/E0ykAhT.jpg
20 FPS: http://i.imgur.com/zckioEp.jpg

**Comment by badock** [comment]
Hello,

How did you manage to display the white lines composing objects, as in the screenshots?

**Comment by ensiform** [comment]
`r_showtris 1`

You might need `developer` set to `1` first to use it.

**Comment by badock** [comment]
I took a look into this issue, and it appears to me that when someone fire with a blaster, the blaster's FX effect will be rendered in each of the next frames (until the blaster shot hits something). 

I saw is that a blaster's FX effect seems to have a lifetime that can be superior to the frame's lifetime (especially if the FPS is high), causing several effects of a same blaster shot (belonging to different frames at different times) to be superposed.

If I am not mistaken, a solution could be to make the FX scheduler check if a blaster shot is already being rendered before creating a new blaster effect.

**Comment by badock** [comment]
Hello,

I wrote a fix for this issue. In situation of high framerate, blaster shots are giving the impression of being rendered several times. It is due to the fact that the FX corresponding has a life time which is constant.

Basically the fix adjusts the lifetime of a blaster shot FX effect, so that it looks like what is would be rendered at 60 fps.

here is a link to the fix:
https://github.com/badock/OpenJK/commit/0c282ca74bbed8bf889adc42642accb6e91b757e

do not hesitate to make comments.

Regards,

Badock

**Comment by badock** [comment]
Yes, the fix is currently singleplayer only. 

If this fix proposal is accepted, we would have to apply the same modification in codemp/client/FxScheduler.cpp .

**Comment by ensiform** [comment]
I am hesitant to implement because JK2 does not do this and overall the code is very similar but does not suffer the problem and same effect files/calling functions.

**Comment by badock** [comment]
I noticed a difference regarding the duration of effects between JK2 and JKA: 
[JK2](https://github.com/JACoders/OpenJK/blob/712627afab20aaa772d062a8fc8aeb5d3efa6152/codeJK2/cgame/FxTemplate.cpp#L51) vs [JKA](https://github.com/JACoders/OpenJK/blob/712627afab20aaa772d062a8fc8aeb5d3efa6152/code/cgame/FxTemplate.cpp#L53).

Using the following value in JKA:
```
mLife.SetRange( 1.0f, 1.0f );
```

seems to solve the issue with the E11 blaster (imperial blaster), but does not solve the issue with the Bryar pistol. I suspect that the effect corresponding to the E11 blaster uses the default mLife value, while the effect of the Bryar pistol seems to override the default value.

Do you know where those mLife values are defined?

NB: I could also update my fix and use a constant fx_duration value.

```
float fx_duration = 1.0f;
```

It would correspond to what is rendered in JK2:

**Comment by ensiform** [comment]
There is likely a reason they made life range 50-50 in JKA though. There is extra primitives and I think the ability to call other emitters is new.

**Comment by badock** [comment]
Are the efx files the same between JediAcademy and JediOutcast?

**Comment by Grab11** [comment]
Yes

**Comment by badock** [comment]
Well, in fact there are differences in some EFX files : I compared the EFX file for the Bryar pistol's shot that is included in JediAcademy and JediOutcast, and while the [JediAcademy version](https://gist.github.com/badock/993bed6610444e9645149fdd095bdc85) provides a "life" value for each of the component of the effect, the [JediOutcast version](https://gist.github.com/badock/413b8e93074f2290347b024b308d1fee) does not provide any value for mLife.

I suppose than in JediOutcast the Bryar pistol's shot was inheriting the default mLife value (1.0f) which was smaller than the default mLife value of JediAcademy. 

It may explain why reverting to the JediOutcast default mLife value:

```
mLife.SetRange( 1.0f, 1.0f );
```
was fixing most of the weapons, but not the bryar pistol. 

It seems that it will be hard to implement a "clean fix" to the bug reported in this issue.

**Comment by ensiform** [comment]
Doesn't mLife variable come from the life key in the file?

**Comment by badock** [comment]
(My apologies, I have confused mLife and life key in some parts of my previous comment, and also I uploaded the wrong code in the JediAcademy's Gist. I have correct my previous comment!).

You are right, the mLife variable comes from the life key in EFX files.

I noticed that In JediOutcast, the EFX file for the Bryar shot does not contain a life key (c.f. [gist](https://gist.github.com/badock/413b8e93074f2290347b024b308d1fee)), thus when the Bryar FX effect is created, its mLife value is inherited from the default value (1.0f). In JediAcademy, the EFX file contains a life key ([here](https://gist.github.com/badock/993bed6610444e9645149fdd095bdc85#file-ja-bryar-shot-efx-L9)), thus the mLife value will be different from the JediOutcast one. As in JediAcademy the mLife value is quite big here (10), it causes:
- an "after-image" effect, which gives the impression that FX effects are rendered several times. 
- the rendering result defends of the number of frame per second : more FPS => shots seems larger.

I can see two ways to implement a fix:
 1. Ignore the duration of FX effects for weapon shots : they would have a constant duration (for example 1.0).
 2. Make the rendering of weapons shots independent of the FPS by adapting the duration of FX effects : 60fps may be the reference for the rendering.

What would you recommend?

**Comment by deathsythe47** [comment]
OpenJK has used hacks to improve Raven assets in the past, how about hacking only projectile effects to use 1.0 duration? Maybe check the name of the effect before its primitives are drawn.

**Comment by badock** [comment]
I quite agree. Would this fix (duration of 1.0) take place when parsing EFX files (mLife set to 1.0) or rather in the FXScheduler (setting duration of FX effect to 1.0)?

**Comment by ensiform** [comment]
We haven't really done so or shipped any assets though.  The workarounds for the broken shaders were only because we corrected an issue and caused them.  And I don't see any easy way to kludge specific efx primitives in certain efx files if they match a certain filename AND hash/pattern of text to be replacing.

**Comment by deathsythe47** [comment]
By hacked assets I was referring to 93bffc438c501c386b889c755ab6beb29bdd4000

How about when the efx file is parsed, if the efx filename matches any of the known projectile filenames, we force all of its primitives to have mLife of 1.0. No need to check file hashes especially since many custom weapon effects mods likely copied the `life` parameter from Raven's and thus also have the bug. For example I just checked the bryar shot from https://jkhub.org/files/file/2321-ultimateweapons/ and it has `life` all over the place.

**Comment by ensiform** [comment]
Said hack only exists because those assets aren't changeable for real without something like jk2mv's patching system since the menus files aren't GPL.

I'm not sure it will work that way though as far as all prims based on this file.... I'll have to look at the code when I'm working on a few things later.

**Comment by deathsythe47** [comment]
Working for multiplayer in my fork https://github.com/jkanewmod/NewJK/commit/f8602fa774aab84b08ab9096bab93eb8862dad80

**Comment by ensiform** [comment]
Using a variant of your fix @deathsythe47 but it kinda makes the conc rifle's projectile not very visible.  Also affects rocket smoke.

**Comment by deathsythe47** [comment]
You are right. How about instead of checking efx files against a list of all known projectile filenames, just simply restore `mLife.SetRange( 1.0f, 1.0f );` in `CPrimitiveTemplate::CPrimitiveTemplate()` but for use a little hack for bryar only: check to see if `life` is exactly 10 for bryar projectile efx. If it's 10 then we know it's either the base file, or a custom effect that was copied from the base file, so we can force it to 1.0. Otherwise if `life` is something other than 10, it could be some custom effect where the creator *did* intend to do it, e.g. using `life 200 400` for a smoke trail like with rockets, in which case we use that value.

**Comment by Razish** [comment]
File-specific hacks do not belong.
The assets carried over from JK2 should be fixed/adjusted to accommodate for the default life value changing.
Changing the default life value globally is undesirable, new JKA effects were designed with this in mind.

Would be interesting to see if the results in [grab's comment](https://github.com/JACoders/OpenJK/issues/908#issuecomment-279009403) occurred on JK2. Then we still have an issue.

Voting to close.
